mlir/lib/Dialect/Linalg/Transforms/Promotion.cpp

```diff
@@ -174,6 +177,11 @@ LinalgOpInstancePromotionOptions::LinalgOpInstancePromotionOptions(
     Operation *op = opOperand.get().getDefiningOp();
     if (auto sv = dyn_cast_or_null<memref::SubViewOp>(op)) {
       subViews[operandNumber] = sv;
+      // In case of linalg generic, copy in only if subview is used in linalg
+      // payload.
+      if (!isa<linalg::GenericOp>(linalgOp) ||
```

Comment:
why is tied to a linalg.generic? Can the optimization works for any LinalgOp?

Reply:
@chelini In case of `linalg.generic`, it is easy to know if an input/output operand are being used by checking if the operand is being used in the payload before except writing to it (which is what I check `payloadUsesValueFromOperand`).
If case of any `LinalgOp`, I am not sure it is triavial since some operations doesn't have a payload (i.e. linalg.copy).

One think we can do it to add an interface for each LinalgOp to implement and check if output operand (init) might be used before assignment.

Reply:
`payloadUsesValueFromOperand` is an interface method of `LinalgStructuredInterface`, and the copy operation implements `LinalgStructuredInterface` too. Looking at `LinalgNamedStructuredOps.yamlgen.td` (you find it in the build dir) CopyOp "derives" from `LinalgStructuredBase_Op`, which implements `LinalgStructuredInterface`, thus `payloadUsesValueFromOperand` is well defined also for `CopyOp` too.

Reply:
You are right about `CopyOp` has `payloadUsesValueFromOperand `, but for that example the `ins` operand needs promotation altough `payloadUsesValueFromOperand ` will return false for it since `linalg.copy` has not payload